NLS eReader Zoomax braille driver

florin-trutiu opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Closes #15863

### Summary of the issue:
This is the NVDA driver for the NLS eReader Zoomax braille display.
It supports both USB and Bluetooth automatic detection.

### Description of user facing changes
With this driver the user can use directly the NLS eReader Zoomax display without the need to manually install it as an addon.

### Description of development approach
The driver is similar with the existing braille display drivers for NVDA.

### Testing strategy:

### Known issues with pull request:

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

seanbudd commented:
Thanks for opening this PR @florin-trutiu !

- Please fill out the PR template
- Please remove this line from the userGuide: https://github.com/nvaccess/nvda/blob/c17b4230fde8a1d9cb811181bc4f4e8c19dc8234/user_docs/en/userGuide.md#L4284
- Please add a changelog entry in changes.md noting that support for this braille device has been added

seanbudd commented:
pre-commit.ci autofix

florin-trutiu commented:
> Thanks for opening this PR @florin-trutiu !
> 
> * Please fill out the PR template
> * Please remove this line from the userGuide: https://github.com/nvaccess/nvda/blob/c17b4230fde8a1d9cb811181bc4f4e8c19dc8234/user_docs/en/userGuide.md#L4284
> * Please add a changelog entry in changes.md noting that support for this braille device has been added

Removed the required line.
I noticed that the line was in that section because of a confusion, there are two NLS eReader braille displays. one is manufactured by Humanware and one by Zoomax.
So to avoid confusion I appended the Humanware name to the NLS eReader device mentioned in the Humanware displays section.

In the user guide I have now added a section for the NLS eReader Zoomax braille display.

CyrilleB79 commented:
@florin-trutiu, you have added the template, but for completeness, could you please writing the required information in its sections rather than writing before it? Thanks!

SaschaCowley commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+		:type command: bytes
+		:type arg: bytes | bool | int
+		"""
+		typeErrorString = "Expected param '{}' to be of type '{}', got '{}'"
```
Could you change this to use named rather than positional fields?
```suggestion
		TYPE_ERROR_TEMPLATE = "Expected parameter '{paramName}' to be of type '{expectedType}', got '{actualType}'"
```

SaschaCowley commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+		if not isinstance(command, bytes):
+			raise TypeError(typeErrorString.format("command", "bytes", type(command).__name__))
+
+		arg = arg.replace(ESCAPE, ESCAPE * 2)
```
Can you please explain why you are doing this?

SaschaCowley commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+
+	def _onReceive(self, data: bytes):
+		if data != ESCAPE:
+			log.debugWarning("Ignoring byte before escape: %r" % data)
```
Please use an f-string instead

SaschaCowley commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+			# If not, we wish to know about it, allow decode to raise.
+			self._deviceID = arg.decode("latin-1", errors="strict")
+		elif command in KEY_NAMES:
+			arg = sum(byte << offset * 8 for offset, byte in enumerate(arg))
```
Why not
```suggestion
			arg = int.from_bytes(reversed(arg))
```

SaschaCowley commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+			self._deviceID = arg.decode("latin-1", errors="strict")
+		elif command in KEY_NAMES:
+			arg = sum(byte << offset * 8 for offset, byte in enumerate(arg))
+			if arg < self._keysDown.get(command, 0):
+				# Release.
+				if not self._ignoreKeyReleases:
+					# The first key released executes the key combination.
+					try:
+						inputCore.manager.executeGesture(InputGesture(self._deviceID, self._keysDown))
+					except inputCore.NoInputGestureAction:
+						pass
+					# Any further releases are just the rest of the keys in the combination being released,
+					# so they should be ignored.
+					self._ignoreKeyReleases = True
+			else:
+				# Press.
+				# This begins a new key combination.
+				self._ignoreKeyReleases = False
+			if arg > 0:
+				self._keysDown[command] = arg
+			elif command in self._keysDown:
+				# All keys in this group have been released.
+				# #3541: Remove this group so it doesn't count as a group with keys down.
+				del self._keysDown[command]
```
I'm having a lot of trouble following what's going on here. Could you please comment a bit more on what's going on?

SaschaCowley commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+
+		self.keyNames = names = []
+		for group, groupKeysDown in keysDown.items():
+			if group == LOC_BRAILLE_KEYS and len(keysDown) == 1 and not groupKeysDown & 0xF8:
```
Why `0xf8`?

SaschaCowley reviewed (commented):
Thanks for submitting this, @florin-trutiu! A lot of questions and stylistic questions on this to begin with

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -27,6 +27,7 @@ To use this feature, "allow NVDA to control the volume of other applications" mu
 * Automatic language switching is now supported when using Microsoft Speech API version 5 (SAPI5) and Microsoft Speech Platform voices. (#17146, @gexgd0419)
 * NVDA can now be configured to speak the current line or paragraph when navigating with braille navigation keys. (#17053, @nvdaes)
 * In Word, the selection update is now reported when using Word commands to extend or reduce the selection (`f8` or `shift+f8`). (#3293, @CyrilleB79)
+* NLS eReader Zoomax driver has been added. Previously it was available only as a separate addon.
```
```suggestion
* Support for the NLS eReader Zoomax braille display has been added. (#17509, @florin-trutiu)
```

SaschaCowley commented on user_docs/en/userGuide.md:
```diff
@@ -5146,6 +5146,34 @@ Please see the display's documentation for descriptions of where these keys can
+
+The NLS eReader Zoomax device supports USB or bluetooth connections.
+The Windows 10 and Windows 11 operating systems will automatically detect and install the necessary drivers for this display.
+For computers where the Internet connection is disabled or not available, for the USB connection, a driver can be download from the USB to serial CH340 chip manufacturer  (https://www.wch-ic.com/downloads/CH341SER_EXE.html)
```
```suggestion
For computers where the Internet connection is disabled or not available, you can manually [download and install the USB to serial CH340 chip driver](https://www.wch-ic.com/downloads/CH341SER_EXE.html) to support this display over USB.
```

florin-trutiu commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+CONNECT_RETRIES = 5
+TIMEOUT_BETWEEN_RETRIES = 2
+
+ESCAPE = b"\x1b"
```
@SaschaCowley 
The nlseReaderZoomax.py driver uses code from the baum.py driver that is already part of the NVDA source code.
This is done so because the NLS eReader Zoomax has a protocol similar with the one used by the Baum Braille displays.
So the plan was to provide a separate driver file but to not change the functionality that was already working and approved by the NVDA team.

florin-trutiu replied:
I discussed with my manager and I will make the also the rest of the changes your requested but it will take more time as I have a high priority project that needs to be done first.

florin-trutiu commented on source/brailleDisplayDrivers/nlseReaderZoomax.py:
```diff
@@ -0,0 +1,295 @@
+		driverRegistrar.addUsbDevices(
+			bdDetect.DeviceType.SERIAL,
+			{
+				"VID_1A86&PID_7523",  # CH340
```
@SaschaCowley 
CH340 is an USB to serial chip used in the device, and those are it's VID and PID values.

seanbudd commented:
@florin-trutiu - Is this ready for re-review? I noticed you pushed some changes in January, but this was never marked as ready for review again

Adriani90 commented:
it seems @florin-trutiu is not quite active anymore on this repo. Should we mark this as abandoned? Or should we ask within the community whether someone else wants to take it over?

florin-trutiu commented:
> it seems @florin-trutiu is not quite active anymore on this repo. Should we mark this as abandoned? Or should we ask within the community whether someone else wants to take it over?
@seanbudd 
The NLS eReader Zoomax braille driver is/was similar with the rest of the Braille display drivers already present in the NVDA sources.
The problem appeared when I was requested by @SaschaCowley to make multiple changes to the driver, although, like I mentioned, the driver was similar with the rest of the drivers and those drivers work and do not have the changes I was requested to make.
At that time I replied to @SaschaCowley and told him that I talked with my manager and that I will make the changes, but as it required major changes, I could not find the time to do it..
I will discuss again the issue with those in charge to see if I can move it higher in my priorities list.
